### PR TITLE
improve variantcontext.* javadoc

### PR DESCRIPTION
**Pull request by lindenb**
### Description

* I've added a few sentences in the javadoc to highlight what is happening under the hood of some functions
* I've added a few 'final' keyworkds
* added the method 'hasFilter' in VariantContext and CommonInfo
* added some tests for CommonINfo
* fix getAttributeAsList where the array is a int[] or a double[]

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by droazen** [comment]
@ldgauthier Would you be able to review this one?

**Comment by ldgauthier** [comment]
Can do.  I should be able to get around to it by the end of next week.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 1]
```diff
@@ -151,7 +173,8 @@ public boolean hasLog10PError() {
      */
     public double getPhredScaledQual() { return (getLog10PError() * -10) + 0.0; }
 
-    public void setLog10PError(double log10PError) {
+    /** set the Phred scaled quality score */
```
Comment (ldgauthier) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Is this value Phred-scaled at this point?  It will become the QUAL (in the writer?), but I thought it's just stored as log10prob.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 2]
```diff
@@ -90,38 +91,58 @@ public void setName(String name) {
...
     public Set<String> getFiltersMaybeNull() {
         return filters;
     }
 
+    /** @return an unmodifiable Set of filters. Can be empty by never null */
     public Set<String> getFilters() {
         return filters == null ? NO_FILTERS : Collections.unmodifiableSet(filters);
     }
 
+    /** @return true if no filter has been defined  <code>(getFiltersMaybeNull()==null)</code> */
```
Comment (ldgauthier) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: It would be nice to note somewhere that in the parsing in the VCF codec if a site is unfiltered (represented by VCFConstants.UNFILTERED=".") then the filters list is set to null and that if the VC passes filters ("PASS"), its common info gets an empty Set<String>.

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 3]
```diff
@@ -721,33 +721,50 @@ public String getID() {
     // get routines to access context info fields
     //
     // ---------------------------------------------------------------------------------------------------------
+    /** shortcut to {@link CommonInfo#getName()} */
```
Comment (ldgauthier) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: I would rather have the javadoc in VariantContext.java instead of CommonInfo.  Or both.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 4]
```diff
@@ -247,41 +284,122 @@ public Object getAttribute(String key, Object defaultValue) {
...
         as a Collections.singletonList if there is only one value */
     @SuppressWarnings("unchecked")
-    public List<Object> getAttributeAsList(String key) {
-        Object o = getAttribute(key);
+    public List<Object> getAttributeAsList(final String key) {
+        final Object o = getAttribute(key);
         if ( o == null ) return Collections.emptyList();
-        if ( o instanceof List ) return (List<Object>)o;
-        if ( o.getClass().isArray() ) return Arrays.asList((Object[])o);
+        else if ( o instanceof List ) return (List<Object>)o;
```
Comment (ldgauthier) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: I appreciate that this is a problem, but I hate to add more instanceof calls.  I guess that's an implicit vote for the proposal by @tfenne in #520

**Comment by lindenb** [comment]
@ldgauthier back to you and thank you for the review.  I've copied the javadoc from CommonInfo to the shotcuts in VariantContext.
Regarding  your other comments, there is not much I can do  isnt't it ? :-) ( setLog10PError , 'instanceof' calls...)

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 5]
```diff
@@ -721,45 +721,122 @@ public String getID() {
...
-    /** shortcut to {@link CommonInfo#hasLog10PError()} */
+    
+    /** shortcut to {@link CommonInfo#hasLog10PError()}
+     * @return true if log10-based error estimate has been set 
+     */
     public boolean hasLog10PError()             { return commonInfo.hasLog10PError(); }
-    /** shortcut to {@link CommonInfo#getLog10PError()} */
+    
+    /** shortcut to {@link CommonInfo#getLog10PError()}
+     * @return the -1 * log10-based error estimate
```
Comment (ldgauthier) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: Where does the -1 come from?

**Comment by lindenb** [comment]
@ldgauthier 
> Where does the -1 come from?

it's here: https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/variant/variantcontext/CommonInfo.java#L139

**Comment by ldgauthier** [comment]
I don't see where in the code the sign is flipped, except in getPhredScaledQual, which is a -10. Log10PError should be negative and that method just returns it. @droazen is it possible you made a mistake in that original comment?

**Comment by ldgauthier** [comment]
I'm still hesitant to add more instanceof calls, but I guess it's just a question of doing them here or in whatever package is working with the VCF.  This way seems safer. @droazen Do you have any objections? Ultimately, it would be nice to be able to use the type declared in the header for each info field attribute, but that's a pretty significant change.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 6]
```diff
@@ -90,38 +91,58 @@ public void setName(String name) {
     //
     // ---------------------------------------------------------------------------------------------------------
 
+    /** @return a <b>modifiable</b> Set of filters. Can be  null. All changes in this set will be reflected in the CommonInfo */
```
Comment (nh13) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: extra space `be__null`

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 7]
```diff
@@ -90,38 +91,58 @@ public void setName(String name) {
...
     public Set<String> getFiltersMaybeNull() {
         return filters;
     }
 
+    /** @return an unmodifiable Set of filters. Can be empty by never null */
```
Comment (nh13) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: typo (`by`)

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 8]
```diff
@@ -247,41 +284,122 @@ public Object getAttribute(String key, Object defaultValue) {
...
+     * if given key is not found the defaultValue is returned.
+     * if the value is as an Integer, this value is returned.
+     * if the value is a String, the value of <code>Double.parseDouble((String)x)</code> is returned.
+     * If the value is not a Double, a String or an Integer, an exception is thrown
+     * 
+     * @param key the attribute key
+     * @param defaultValue the default value
+     * @return the attribute as a double
+     */
+    public double getAttributeAsDouble(final String key, final double defaultValue) {
```
Comment (nh13) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: I am so confused why in `VariantContext.java` we have these `getAttributeAs*` methods but in `Genotype.java` the similar methods are deprecated and implemented differently.  Likely this is for a different PR, but perhaps with some consensus a new PR could emerge?

**Comment by SHuang-Broad** [comment]
To add a somewhat relevant comment, if one looks at the first constructor in the `Allele` classes--`protected Allele(final byte[] bases, final boolean isRef)`--it says "null alleles are no longer allowed".

But the documentation for the public method `public static Allele create(final byte[] bases, final boolean isRef)` says "If bases == '-', a Null allele is created", which seems like an outdated doc, because the code in the `create` method doesn't actually check for the case of dash valued bases.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java (hunk 8 above)
Comment (lbergelson) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: I recently ran into that and was also confused.

**Comment by yfarjoun** [comment]
@lindenb This PR claims to be a documentation PR, but has many code changes, not all of which are trivial. Could you please split it into two separate PRs one with the doc changes and another one with the code changes together with appropriate tests for the new functionality?

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java (hunk 8 above)
Comment (lindenb) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: @lbergelson what do you mean ? do you mean you don't understand the original javadoc or you don't understand my changes ?

Comment (lbergelson) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Oh, I was agreeing with @nh13  that we had weird api differences between `VariantContext` and `Genotype` that should probably be looked at in a different PR.  Not a comment on the doc at all.

**Comment by lindenb** [comment]
@yfarjoun I've removed the code changes. I've only kept the 'final' keywords. Shall I remove them too ?

**Comment by yfarjoun** [comment]
I think it's fine to leave the 'final's. thanks.

On Wed, Jun 29, 2016 at 5:57 AM, Pierre Lindenbaum <notifications@github.com

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java (hunk 2 above)
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: This comment is backwards -- the method returns true if `filters != null`, not if "no filter has been defined"

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 9]
```diff
@@ -90,27 +91,43 @@ public void setName(String name) {
...
     public Set<String> getFilters() {
         return filters == null ? NO_FILTERS : Collections.unmodifiableSet(filters);
     }
 
+    /** @return true if no filter has been defined  <code>(getFiltersMaybeNull()==null)</code> */
     public boolean filtersWereApplied() {
         return filters != null;
     }
 
+    /** @return true if any filter been defined  <code>(!filters.isEmpty())</code> */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Better: "true if filters were applied and the VC did not pass filters"

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 10]
```diff
@@ -90,27 +91,43 @@ public void setName(String name) {
...
     public boolean isFiltered() {
         return filters == null ? false : !filters.isEmpty();
     }
 
+    /**
+     * @param filter the filter ID 
+     * @return true if the filters contains 'filter' */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Document that this method won't work for `PASS`, since it's not included in the set of filters

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 11]
```diff
@@ -151,7 +173,8 @@ public boolean hasLog10PError() {
      */
     public double getPhredScaledQual() { return (getLog10PError() * -10) + 0.0; }
 
-    public void setLog10PError(double log10PError) {
+    /** set the Phred scaled quality score */
+    public void setLog10PError(final double log10PError) {
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: The comment for this method is incorrect -- `log10PError` is not Phred-scaled, as @ldgauthier suspected. You can see this by looking at the implementation of the `getPhredScaledQual()` method above.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 12]
```diff
@@ -166,29 +189,33 @@ public void setLog10PError(double log10PError) {
...
      */
     public Map<String, Object> getAttributes() {
         return Collections.unmodifiableMap(attributes);
     }
 
     // todo -- define common attributes as enum
-
-    public void setAttributes(Map<String, ?> map) {
+    /** clear all attributes, and set 'map' as attributes */
+    public void setAttributes(final Map<String, ?> map) {
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Document whether subsequent changes to the `map` passed in will be reflected in this record (they won't).

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 13]
```diff
@@ -166,29 +189,33 @@ public void setLog10PError(double log10PError) {
...
         clearAttributes();
         putAttributes(map);
     }
 
-    public void putAttribute(String key, Object value) {
+    /** shortcut of <code>putAttribute(key, value, false)</code> */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Add proper javadoc for this method (and for the other methods where you've written "shortcut of x")

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 14]
```diff
@@ -166,29 +189,33 @@ public void setLog10PError(double log10PError) {
...
+    public void putAttribute(final String key, final Object value) {
         putAttribute(key, value, false);
     }
 
-    public void putAttribute(String key, Object value, boolean allowOverwrites) {
+    /** insert a new attribute, raises an exception if allowOverwrites==true and the attribute exists */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: This isn't right -- the exception is raised if allowOverwrites is **false** and the attribute exists.

Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: Also, use a proper javadoc-style `@throws` to document exceptions

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 15]
```diff
@@ -198,31 +225,35 @@ public void putAttribute(String key, Object value, boolean allowOverwrites) {
...
-    public void removeAttribute(String key) {
+    /** removes the attribute identified by key */
+    public void removeAttribute(final String key) {
         if ( attributes == NO_ATTRIBUTES ) // immutable -> mutable
             attributes = new HashMap<String, Object>();
         attributes.remove(key);
     }
 
-    public void putAttributes(Map<String, ?> map) {
+    /** insert the attibutes as map using <code>putAttribute(key,value,false)</code> */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: "as map" -> "in the map"

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 16]
```diff
@@ -198,31 +225,35 @@ public void putAttribute(String key, Object value, boolean allowOverwrites) {
...
-                for ( Map.Entry<String, ?> elt : map.entrySet() ) {
+                for ( final Map.Entry<String, ?> elt : map.entrySet() ) {
                     putAttribute(elt.getKey(), elt.getValue(), false);
                 }
             }
         }
     }
 
-    public boolean hasAttribute(String key) {
+    /** @return true if the key is present */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: key -> attribute

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 17]
```diff
@@ -247,37 +284,78 @@ public Object getAttribute(String key, Object defaultValue) {
...
+     * 
+     * @param key the attribute key
+     * @param defaultValue  the default value
+     * @return the attribute as a String
+     */
+    public String getAttributeAsString(final String key, final String defaultValue) {
+        final Object x = getAttribute(key);
         if ( x == null ) return defaultValue;
         if ( x instanceof String ) return (String)x;
         return String.valueOf(x); // throws an exception if this isn't a string
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: This comment "throws an exception if this isn't a string" isn't right -- `String.valueOf(x)` calls `x.toString()`, regardless of whether x is already a String or not.

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 18]
```diff
@@ -247,37 +284,78 @@ public Object getAttribute(String key, Object defaultValue) {
...
     }
 
-    public int getAttributeAsInt(String key, int defaultValue) {
+    /**
+     * return an attribute as an integer.
+     * if given key is not found the defaultValue is returned.
+     * if the value is a String, the value of <code>Integer.parseInt((String)x)</code> is returned.
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: `parseInt` -> `valueOf`

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 19]
```diff
@@ -247,37 +284,78 @@ public Object getAttribute(String key, Object defaultValue) {
...
         return Integer.valueOf((String)x); // throws an exception if this isn't a string
     }
-
-    public double getAttributeAsDouble(String key, double defaultValue) {
+    
+    /**
+     * return an attribute as a double.
+     * if given key is not found the defaultValue is returned.
+     * if the value is as an Integer, this value is returned.
+     * if the value is a String, the value of <code>Double.parseDouble((String)x)</code> is returned.
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: `parseDouble` -> `valueOf`

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 20]
```diff
@@ -247,37 +284,78 @@ public Object getAttribute(String key, Object defaultValue) {
...
         if ( x instanceof Integer ) return (Integer)x;
         return Integer.valueOf((String)x); // throws an exception if this isn't a string
     }
-
-    public double getAttributeAsDouble(String key, double defaultValue) {
+    
+    /**
+     * return an attribute as a double.
+     * if given key is not found the defaultValue is returned.
+     * if the value is as an Integer, this value is returned.
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: "if the value is an Integer or Double, it is returned as a Double"

File: src/java/htsjdk/variant/variantcontext/CommonInfo.java [hunk 21]
```diff
@@ -247,37 +284,78 @@ public Object getAttribute(String key, Object defaultValue) {
...
         if ( x instanceof Double ) return (Double)x;
         if ( x instanceof Integer ) return (Integer)x;
         return Double.valueOf((String)x); // throws an exception if this isn't a string
     }
 
-    public boolean getAttributeAsBoolean(String key, boolean defaultValue) {
+    /**
+     * return an attribute as a boolean.
+     * if given key is not found the defaultValue is returned.
+     * if the value is a String, the value of <code>Boolean.parseBoolean((String)x)</code> is returned.
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/CommonInfo.java]: `parseBoolean` -> `valueOf`

File: src/java/htsjdk/variant/variantcontext/Genotype.java [hunk 22]
```diff
@@ -181,6 +181,10 @@ public boolean hasDP() {
     // ---------------------------------------------------------------------------------------------------------
 
     /**
+     * if {@link #getAlleles()} is empty, the type is {@link GenotypeType#UNAVAILABLE}.
+     * if all the  alleles are NoCall,  the type is {@link GenotypeType#NO_CALL}.
+     * if some, but not all, alleles are NoCall,  the type is {@link GenotypeType#MIXED}.
+     *
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/Genotype.java]: Your doc for this method is missing any description of what happens when **none** of the alleles are no call (ie., you get `GenotypeType.HET`, `GenotypeType.HOM_REF`, or `GenotypeType.HOM_VAR` as appropriate).

File: src/java/htsjdk/variant/variantcontext/Genotype.java [hunk 23]
```diff
@@ -255,16 +259,19 @@ else if ( !allele.equals(observedAllele) )
     public boolean isNoCall() { return getType() == GenotypeType.NO_CALL; }
 
     /**
+     * shortcut of ( {@link #getType()} != NO_CALL AND {@link #getType()} != UNAVAILABLE )
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/Genotype.java]: `NO_CALL` -> `{@link GenotypeType#NO_CALL}`
`UNAVAILABLE` -> `{@link GenotypeType#UNAVAILABLE}`

File: src/java/htsjdk/variant/variantcontext/Genotype.java [hunk 24]
```diff
@@ -255,16 +259,19 @@ else if ( !allele.equals(observedAllele) )
...
      * @return true if this genotype is comprised of any alleles that are not no-calls (even if some are).
      */
     public boolean isCalled() { return getType() != GenotypeType.NO_CALL && getType() != GenotypeType.UNAVAILABLE; }
 
     /**
+     * shortcut of ( {@link #getType()} == MIXED )
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/Genotype.java]: `MIXED` -> `{@link GenotypeType#MIXED}`

File: src/java/htsjdk/variant/variantcontext/Genotype.java [hunk 25]
```diff
@@ -255,16 +259,19 @@ else if ( !allele.equals(observedAllele) )
...
      * @return true if this genotype is comprised of both calls and no-calls.
      */
     public boolean isMixed() { return getType() == GenotypeType.MIXED; }
 
     /**
+     * shortcut of ( {@link #getType()} != UNAVAILABLE )
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/Genotype.java]: `UNAVAILABLE` -> `{@link GenotypeType#UNAVAILABLE}`

File: src/java/htsjdk/variant/variantcontext/GenotypesContext.java [hunk 26]
```diff
@@ -660,6 +660,7 @@ public boolean containsSample(final String sample) {
         return sampleNameToOffset.containsKey(sample);
     }
 
+    /** return true if all {@link GenotypesContext#getSampleNames() contains all sample */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/GenotypesContext.java]: Wording is awkward here -- "return true if the list of samples in this context as returned by `{@link #getSampleNames()}` contains all samples in the Collection passed in"

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 27]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+     */
     public String getSource()                   { return commonInfo.getName(); }
+    
+    /** shortcut to {@link CommonInfo#getFiltersMaybeNull()} 
+     *  @return a <b>modifiable</b> Set of filters. Can be  null. All changes in this set will be reflected in the CommonInfo 
+     */
     public Set<String> getFiltersMaybeNull()    { return commonInfo.getFiltersMaybeNull(); }
+    
+    /** shortcut to {@link CommonInfo#getFilters()} 
+     * @return an unmodifiable Set of filters. Can be empty by never null
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: by -> but

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 28]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+     */
     public Set<String> getFilters()             { return commonInfo.getFilters(); }
+    
+    /** shortcut to {@link CommonInfo#isFiltered()}
+     * @return true if any filter been defined  <code>(!filters.isEmpty())
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: "true if filters were applied and the VC did not pass filters"

Fix broken `<code>` tag

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 29]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+     */
     public boolean isFiltered()                 { return commonInfo.isFiltered(); }
+    
+    /** shortcut to {@link CommonInfo#isNotFiltered()} */
     public boolean isNotFiltered()              { return commonInfo.isNotFiltered(); }
+    
+    /** shortcut to {@link CommonInfo#filtersWereApplied()}
+     *  @return true if no filter has been defined  <code>(getFiltersMaybeNull()==null)</code>
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: This is backwards -- this returns true if `filters != null`

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 30]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+     * @return an unmodifiable Set of filters. Can be empty by never null
+     */
     public Set<String> getFilters()             { return commonInfo.getFilters(); }
+    
+    /** shortcut to {@link CommonInfo#isFiltered()}
+     * @return true if any filter been defined  <code>(!filters.isEmpty())
+     */
     public boolean isFiltered()                 { return commonInfo.isFiltered(); }
+    
+    /** shortcut to {@link CommonInfo#isNotFiltered()} */
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: "equivalent to ! isFiltered()"

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 31]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+     */
     public boolean filtersWereApplied()         { return commonInfo.filtersWereApplied(); }
+    
+    /** shortcut to {@link CommonInfo#hasLog10PError()}
+     * @return true if log10-based error estimate has been set 
+     */
     public boolean hasLog10PError()             { return commonInfo.hasLog10PError(); }
+    
+    /** shortcut to {@link CommonInfo#getLog10PError()}
+     * @return the -1 * log10-based error estimate
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: This isn't right -- it's just log10(perror), not -1 * log10(perror). You can see this is true because in `commonInfo.getPhredScaledQual()` we multiply `getLog10PError()` by -10.

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 32]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
     public double getPhredScaledQual()          { return commonInfo.getPhredScaledQual(); }
-
+    
+    /** shortcut to {@link CommonInfo#getAttributes()}
+     * @return an unmodifiable Map of attributes
+     */
     public Map<String, Object>  getAttributes() { return commonInfo.getAttributes(); }
+    
+    /** shortcut to {@link CommonInfo#hasAttribute(String)}
+     * @return true if the key is present
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: key -> specified attribute

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 33]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+    /** shortcut to {@link CommonInfo#getAttribute(String, Object)}
+     * @return the attribute value for the given key (or defaultValue if not set)
+     */
     public Object getAttribute(String key, Object defaultValue) {
         return commonInfo.getAttribute(key, defaultValue);
     }
-
+    
+    /** shortcut to {@link CommonInfo#getAttributeAsString(String, String)}
+     * return an attribute as a String.
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: You may want to use the javadoc `@see` tag in these methods to reference the appropriate docs in `CommonInfo`

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 34]
```diff
@@ -721,33 +721,125 @@ public String getID() {
...
+     * if given key is not found the defaultValue is returned.
+     * if the value is not a String <code> String.valueOf(x)</code> is returned.
+     * @return the attribute as a String
+     */
     public String getAttributeAsString(String key, String defaultValue)   { return commonInfo.getAttributeAsString(key, defaultValue); }
+    
+    /** shortcut to {@link CommonInfo#getAttributeAsInt(String, int)}
+     * if given key is not found the defaultValue is returned.
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: All the comments I made in `CommonInfo` for these `getAttributeAs*` methods should be applied here as welll.

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 35]
```diff
@@ -759,17 +851,18 @@ public CommonInfo getCommonInfo() {
...
-        Allele ref = REF;
+        final Allele ref = REF;
         if ( ref == null )
             throw new IllegalStateException("BUG: no reference allele found at " + this);
         return ref;
     }
 
 
     /**
+     * shortcut to <code>{@link #getNAlleles()} == 2</code<
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: Fix broken `</code<` tag

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 36]
```diff
@@ -940,7 +1035,7 @@ public boolean hasGenotypes() {
         return ! genotypes.isEmpty();
     }
 
-    public boolean hasGenotypes(Collection<String> sampleNames) {
+    public boolean hasGenotypes(final Collection<String> sampleNames) {
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: No docs for this one?

File: src/java/htsjdk/variant/variantcontext/VariantContext.java [hunk 37]
```diff
@@ -997,6 +1092,9 @@ public GenotypesContext getGenotypes(Set<String> sampleNames) {
         return getGenotypes().getSampleNames();
     }
 
+    /**
+     * @return the list of all sample names in this context, ordered
```
Comment (droazen) [inline src/java/htsjdk/variant/variantcontext/VariantContext.java]: ordered -> ordered by name

**Comment by droazen** [comment]
Review complete, back to @lindenb for changes.

**Comment by yfarjoun** [comment]
Friendly poke... @lindenb we would be very happy to merge once you respond to the comments.

**Comment by lindenb** [comment]
Thanks & back to you, all.
I've move the comments in VariantContext to  `@see CommonInfo#...`

Thanks
P.

**Comment by lindenb** [comment]
ok, I'm going to close that one for now, this is currently diverging too much from the master.
But I thank the reviewers for their time.
